tortoise/verifying_tortoise.go

```diff
@@ -419,75 +398,65 @@ func (t *turtle) calculateExceptions(
 	if startLayer.Before(types.GetEffectiveGenesis()) {
 		startLayer = types.GetEffectiveGenesis()
 	}
-	for layerID := startLayer; !layerID.After(t.Last); layerID = layerID.Add(1) {
-		logger := logger.WithFields(log.Named("diff_layer_id", layerID))
+	for lid := startLayer; !lid.After(t.Last); lid = lid.Add(1) {
+		logger := logger.WithFields(log.Named("diff_layer_id", lid))
 		logger.Debug("checking input vector diffs")
 
-		layerBlockIds, err := t.getLayerBlocksIDs(ctx, layerID)
+		local, err := t.getLocalOpinion(ctx, lid)
 		if err != nil {
 			if errors.Is(err, database.ErrNotFound) {
 				continue
 			}
-			return nil, fmt.Errorf("layer block IDs: %w", err)
+			return nil, err
 		}
 
 		// helper function for adding diffs
-		addDiffs := func(bid types.BlockID, voteClass string, voteVec vec, diffMap map[types.BlockID]struct{}) {
-			if v, ok := baseBlockOpinion[bid]; !ok || simplifyVote(v) != voteVec {
-				logger.With().Debug("added vote diff",
-					log.Named("diff_block", bid),
-					log.String("diff_class", voteClass),
-					log.Named("opinion", v),
-				)
-				if layerID.Before(baseBlockLayerID) {
-					logger.With().Warning("added exception before base block layer, this block will not be marked good",
-						log.Named("diff_block", bid),
-						log.String("diff_class", voteClass))
+		addDiffs := func(logger log.Log, bid types.BlockID, voteVec vec, diffMap map[types.BlockID]struct{}) {
+			if v, ok := baseopinion[bid]; !ok || simplifyVote(v) != voteVec {
+				logger.With().Debug("added vote diff", log.Named("opinion", v))
+				if lid.Before(baselid) {
+					logger.With().Warning("added exception before base block layer, this block will not be marked good")
 				}
 				diffMap[bid] = struct{}{}
 			}
 		}
 
-		// get local opinion for layer
-		layerInputVector, err := t.layerOpinionVector(ctx, layerID)
-		if err != nil {
-			// an error here signifies a real database failure
-			logger.With().Error(errstrUnableToCalculateLocalOpinion, log.Err(err))
-			return nil, err
-		}
-
-		// otherwise, nil means we should abstain
-		if layerInputVector == nil {
-			// still waiting for Hare results, vote neutral and move on
-			logger.With().Debug("input vector is empty, adding neutral diffs", log.Err(err))
-			for _, b := range layerBlockIds {
-				addDiffs(b, "neutral", abstain, neutralDiff)
+		for bid, opinion := range local {
+			usecoinflip := opinion == abstain && lid.Before(t.layerCutoff())
+			if usecoinflip {
+				coin, exist := t.bdp.GetCoinflip(ctx, votinglid)
```

Comment:
are you getting votinglid.Sub(1)? the log below says so

Reply:
no, sorry the log should be adjusted. in current implementation the value should be last layer that tortoise received

Reply:
what's your consideration for not changing it to Sub(1) here?

Reply:
for creating a block we need to use weakcoin from last completed hare `prepare` round. we update Last in tortoise when hare sends a notification, therefore votinglid should be t.Last and not t.Last - 1. 

this doesn't change what is written in the issue that you created, if we will use layer clock in tortoise then we need to use t.Last - 1 here.

Reply:
> therefore votinglid should be t.Last and not t.Last - 1

my confusion is, we should be generating block for the current ticked layer (say `N`). 
so in the happy flow, votinglid should be `t.Last+1` on line 280 and here we use `votinglid.Sub(1)`. i think that's more consistent with the design discussion.

but in the unhappy flow, `t.Last` is the most recent layer hare terminated, but not necessarily `N-1`. `t.Last` can be layer `N-2` while `N-1` hasn't terminated. in this case, i think it's prolly good enough to use coinflip in `t.Last` (`N-2`) as the situation is the same for all nodes.

otherwise we need to change hare to record coinflip when hare started instead of terminated.

Reply:
yes, this is my understanding as well. right now we use latest available that in happy case is CURRENT - 1, sometimes CURRENT - 2. while working on https://github.com/spacemeshos/go-spacemesh/issues/2921 weakcoin saving needs to be moved after preround and then we can always use CURRENT - 1 based on layer clock.